### ISSUE TITLE
[BUG] Critical dependency: the request of a dependency is an expression after upgrade to 7.0.0

### ISSUE THREAD

**Issue by prymitive** [npm/node-semver#299]
# What / Why

I'm getting an error when building up webpack bundle and using semver 7.0.0:

```
Creating an optimized production build...
Compiled with warnings.

./node_modules/semver/index.js
Critical dependency: the request of a dependency is an expression
```

## When

Started with the upgrade from 6.3.0 -> 7.0.0

## Where

https://github.com/prymitive/karma/pull/1256

### Current Behavior

* n/a

### Steps to Reproduce

* n/a

### Expected Behavior

* n/a

## Who

* n/a

## References

I do see some issues that looks similiar on the webpack project:
webpack/webpack-dev-server#212
webpack/webpack#196

**Comment by isaacs** [comment]
Yeah, it uses require() with an argument that isn't a string literal.  That's intentional.  That's why it's a breaking change, because it changes how the module loads.

I'd be open to adding a `semver-preload.js` or something that loads everything inline, but you're still gonna have to change how you pull it in.

If you're only using certain functionality, you can also just pull in that specific module, like `const semverParse = require('semver/functions/parse')` to avoid packing the whole module, which is probably better.

**Comment by isaacs** [comment]
Yeah, do `require('semver/preload')` and it should work for you on v7.1.0.

**Comment by wartab** [comment]
That does not look like good design. 

Version 7 was released with tree-shaking in mind, but I believe that point got entirely missed. Semver confuses tree-shaking and deep imports.

Proper tree-shaking does not need deep imports from modules to operate correctly. If you properly use exports, tree-shaking will remove all unused exports and there will be no need for the hacky imports we are forced to do.

The entire point of tree-shaking is to not use deep imports for smaller bundle sizes. It's also a huge time sink to prorgram imports manually for every function you need. It was already a somewhat decent time sink before, since semver does not use TypeScript definitions. Using TypeScript for semver would probably solve all these issues at once, but I assume that's out of the question.

**Comment by isaacs** [comment]
@wartab I'm sorry that you are not happy with the current design, but it is optimized for more than just your use case.

It sounds like you can get what you want by doing `require('semver/preload')`.  It's not a priority to rewrite this module in a different JS dialect at this time.  There's plenty of room on the registry for you to do so.

**Comment by wartab** [comment]
I was just pointing out what is explicitly mentioned in your changelog (tree-shaking), while not knowing any tool that would actually properly be able to perform tree-shaking with such a structure. It sure will help reduce the amount of loaded files at runtime if they are all present, but it won't help tree-shaking whatsoever.
We won't reinvent the wheel and are in the progress of adapting to your changes, even if I think that if you are jumping into ES6, might as well do it completely.

**Comment by likezero** [comment]
i meet the same error

**Comment by prymitive** [comment]
I’ve changed my imports to pull in specific functions as recommended and it fixed the problem for me with 7.0+
